src/mavlink-router/mainloop.cpp

```diff
@@ -294,17 +294,22 @@ void Mainloop::loop()
             }
             if (events[i].events & EPOLLERR) {
                 remove_fd(p->fd);
-
-                if (g_endpoints[i]->reopen() >= 0) {
-                    log_error("poll error for fd %i, reopening it", p->fd);
-                    add_fd(g_endpoints[i]->fd, g_endpoints[i], EPOLLIN);
-                } else {
+                bool succeeded = false;
+                for (int retries = 1; retries <= 10; retries++) {
+                    log_error("poll error for fd %i, reopening it. Retry: %d", p->fd, retries);
+                    if (g_endpoints[i]->reopen() >= 0) {
+                        add_fd(g_endpoints[i]->fd, g_endpoints[i], EPOLLIN);
+                        succeeded = true;
+                        break;
+                    }
+                    std::this_thread::sleep_for(std::chrono::milliseconds(222));
```

Comment:
there shouldn't be any sleep in the main loop. If you actually need a huge sleep like this then you should actually schedule a callback, not a 222msec sleep.  Rebember this is single-threaded.

Reply:
Before adding this sleep mavlink router exited on error condition and was restarted by supervisor, the delay was much longer. This change tries to avoid the exit by reopening serial port after this long sleep.

Reply:
Even then, there shouldn't be any sleep in the mainloop. Exiting was considered an acceptable behavior as safety measure: we don't really know what happened.

I'm ok with adding a configuration for try reopening the device rather than exiting. But this shouldn't be done with a sleep but rather with a timer (see how a timer is used in the logging)

Reply:
If I use a timer instead of sleep then the main loop will busy loop and use 100% of CPU time. It is designed this way - whenever there is an error it is busy looping. I don't see how to fix this without sleeping on errors.